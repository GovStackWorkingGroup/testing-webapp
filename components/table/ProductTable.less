@import '../../styles/Variables.less';
@import '../../styles/Extends.less';

.table {
  display: flex;
  flex-direction: column;
}

.table-body {
  width: 70vw;
}

.table-results-count {
  height: 52px;
  display: flex;
  align-items: center;
  gap: 5px;
}

.table-results-count > p {
  font-size: 12px;
  color: @text-secondary;
}

.product-table-row-section:extend(.table-row) {
  max-height: 60vh;
  overflow-y: scroll;
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 1px;
  background-color: @white;
}

.product-table-header:extend(.table-header-text) {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  grid-template-rows: 48px;
  border: 1px solid #dfdfea;
  border-radius: 8px 8px 0px 0px;
  align-items: center;
  position: sticky;
}

.product-table-row {
  width: calc(100% + 15px);
  height: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  border-bottom: 1px solid #dfdfea;
  align-items: center;
}

.product-table-row:hover {
  border-width: 0px 1px 1px 1px;
  border-style: solid;
  border-color: #dfdfea;
  box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.25);
}

.product-table-row-opened {
  background-color: @blue-d-1;
}

.details-arrow {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  font-size: 24px;
  padding-right: 10px;
}

.table-bb-section {
  display: flex;
  align-items: center;
}
.table-bb-section > div:first-of-type:extend(.flex-row) {
  width: 100%;
  align-items: center;
}

.table-bb-image {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  height: 48px;
  overflow: hidden;
  gap: 10px;
}

.bb-image {
  padding-top: 15px;
  padding-bottom: 15px;
  filter: invert(30%) sepia(96%) saturate(2946%) hue-rotate(221deg)
    brightness(92%) contrast(94%);
}

.overflow-count {
  display: flex;
  align-items: center;
  height: 48px;
  background-color: @blue-10;
  width: 24px;
  height: 24px;
  border-radius: 50%;
}

.overflow-count > p {
  margin: auto;
  color: @accents-and-btn;
}

.product-last-update {
  color: @text-secondary;
}

.data-loader-message {
  text-align: center;
}

.table-error-handling:extend(.table-row) {
  display: flex;
  width: 100%;
  height: @table-row;
  align-items: center;
  text-align: center;
  justify-content: center;
}

.bb-icon-tooltip {
  background-color: @charts-chips-and-perc;
  text-transform: capitalize;
  font-size: 14px;
  padding: 5px 16px;
}
