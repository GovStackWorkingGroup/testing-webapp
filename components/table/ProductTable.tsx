import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import InfiniteScroll from 'react-infinite-scroll-component';
import { getSoftwaresData, getSoftwareListCount } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const ProductTable = () => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [isErrorFetchingData, setIsErrorFetchingData] = useState<boolean>(false);
  const [isErrorFetchingCount, setIsErrorFetchingCount] = useState<boolean>(false);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>({
    testApp: { field: 'testApp', order: null },
    lastUpdate: { field: 'lastUpdate', order: null },
    overallCompatibility: { field: 'overallCompatibility', order: null }
  });

  console.log(sortBy);

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    fetchData();
  }, [sortBy]);

  const fetchData = async () => {
    const [data, count] = await Promise.all([
      getSoftwaresData(0, sortBy),
      getSoftwareListCount(),
    ]);

    if (data.status) {
      setProductsList(data.data);
    } else {
      setIsErrorFetchingData(true);
    }

    if (count.status) {
      setProductListLength(count.data);
    } else {
      setIsErrorFetchingCount(true);
    }
  };

  const handleLoadMoreData = useCallback(async () => {
    const data = await getSoftwaresData(productsList.length, sortBy);

    if (data.status) {
      setProductsList([...productsList, ...data.data]);
    } else {
      setProductsList([...productsList]);
    }
  }, [productsList]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    setSortBy(tableSortProperties);
  };

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>
          {isErrorFetchingCount ? format('table.no_result_count.message') : productListLength}
        </p>
        <p>
          {isErrorFetchingCount ? null : productListLength === 1
            ? format('table.result.label')
            : format('table.result.plural.label')}
        </p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        {isErrorFetchingData ? (
          <TableErrorHandling />
        ) : (
          <div id='scrollableDiv' className='product-table-row-section'>
            <InfiniteScroll
              scrollableTarget='scrollableDiv'
              dataLength={productsList?.length ?? 0}
              next={handleLoadMoreData}
              hasMore={
                isErrorFetchingData
                  ? false
                  : productListLength > productsList?.length
              }
              loader={<InfiniteScrollCustomLoader />}
              style={{ overflowX: 'hidden' }}
            >
              {productsList.map((product, key) => (
                <ProductTableRow product={product} key={`product-${key}`} />
              ))}
            </InfiniteScroll>
          </div>
        )}
      </div>
    </div>
  );
};

export default ProductTable;
