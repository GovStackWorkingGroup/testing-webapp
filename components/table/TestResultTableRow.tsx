import classNames from 'classnames';
import React from 'react';
import { BuildingBlockEndpointTest } from '../../service/types';

type Props = {
  bbTest: BuildingBlockEndpointTest;
};

const TestResultTableRow = ({ bbTest }: Props) => (
  <div className='test-table-row'>
    {bbTest?.passed ? (
      <div className='test-table-row-status status-passed' data-testid='bb-test-passed'>
        <p>Passed</p>
      </div>
    ) : (
      <div className='test-table-row-status status-failed' data-testid='bb-test-failed'>
        <p>Failed</p>
      </div>
    )}
    <div
      className={classNames({ 'test-table-row-category': bbTest?.method !== '', })}
      data-testid='bb-test-category'>{bbTest?.method === '' ? '-' : bbTest?.method}
    </div>
    <div data-testid='bb-test-endpoint'>
      <p>{bbTest?.endpoint === '' ?  '-' : bbTest?.endpoint}</p>
    </div>
  </div>
);

export default TestResultTableRow;
