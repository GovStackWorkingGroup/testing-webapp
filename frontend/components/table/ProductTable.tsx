import isEqual from 'lodash/isEqual';
import { useCallback, useEffect, useState } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { useIntl } from 'react-intl';
import { getSoftwareListCount, getSoftwaresData } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const defaultSortBy = {
  testApp: { field: 'testApp', order: null },
  lastUpdate: { field: 'lastUpdate', order: null },
  overallCompatibility: { field: 'overallCompatibility', order: null }
} as SoftwaresTableSortByType;

const ProductTable = ({ showAll }: { showAll: boolean }) => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [error, setError] = useState<string | null>(null);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>(defaultSortBy);
  const [loading, setLoading] = useState(true);

  const { formatMessage } = useIntl();
  const format = useCallback((id: string) => formatMessage({ id }), [formatMessage]);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const [data, count] = await Promise.all([
          getSoftwaresData(0, sortBy, 'main', showAll),
          getSoftwareListCount()
        ]);

        if (data.status) {
          setProductsList(data.data);
        }

        if (count.status) {
          setProductListLength(count.data);
        }
      } catch {
        setError('dataFetchFail');
      } finally {
        setLoading(false); // set loading to false once data is fetched or error occurs
      }
    };

    fetchData();
  }, [sortBy]);

  const handleLoadMoreData = useCallback(async () => {
    try {
      const data = await getSoftwaresData(productsList.length, sortBy);

      if (data.status) {
        setProductsList((prev) => [...prev, ...data.data]);
      }
    } catch {
      setError('dataLoadFail');
    }
  }, [productsList, sortBy]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    if (!isEqual(tableSortProperties, sortBy)) {
      setSortBy(tableSortProperties);
    }
  };

  if (error) {
    return <TableErrorHandling />;
  }

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>{productListLength}</p>
        <p>{productListLength === 1 ? format('table.result.label') : format('table.result.plural.label')}</p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        <div id='scrollableDiv' className='product-table-row-section'>
          {loading ? (
            <div className="product-table-row-section">
              <InfiniteScrollCustomLoader />
            </div>
          ) : (
            <>
              <InfiniteScroll
                scrollableTarget='scrollableDiv'
                dataLength={productsList.length}
                next={handleLoadMoreData}
                hasMore={productListLength > productsList.length}
                loader={<InfiniteScrollCustomLoader />}
                style={{ overflowX: 'hidden' }}
              >
                {productsList.map((product, index) => (
                  <ProductTableRow product={product} key={index} />
                ))}
              </InfiniteScroll>
            </>
          )}
        </div>
      </div>
    </div>
  );
};

export default ProductTable;
