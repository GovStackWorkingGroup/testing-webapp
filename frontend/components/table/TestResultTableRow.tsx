import classNames from 'classnames';
import debounce from 'lodash.debounce';
import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import { BuildingBlockEndpointTest } from '../../service/types';
import TextTooltip from '../TextTooltip';

type Props = {
  bbTest: BuildingBlockEndpointTest;
  passCurrentBBTest: () => void;
};

const TestResultTableRow = ({ bbTest, passCurrentBBTest }: Props) => {
  const endpointTextContainer = React.useRef<HTMLDivElement>(null);
  const [isTextOverflowing, setIsTextOverflowing] = useState<boolean>(false);

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  const handleTextOverflowing = useCallback(() => {
    const bbContainer = endpointTextContainer.current;
    if (bbContainer && bbContainer.clientWidth < bbContainer.scrollWidth) {
      setIsTextOverflowing(true);
    } else {
      setIsTextOverflowing(false);
    }
  }, [endpointTextContainer]);

  useEffect(handleTextOverflowing, [handleTextOverflowing]);

  useEffect(
    () =>
      window.addEventListener('resize', debounce(handleTextOverflowing, 20)),
    [handleTextOverflowing]
  );

  return (
    <div className="test-table-row" onClick={passCurrentBBTest}>
      {bbTest?.passed ? (
        <div
          className="test-table-row-status status-passed"
          data-testid="bb-test-passed"
        >
          <p>{format('test_table.passed')}</p>
        </div>
      ) : (
        <div
          className="test-table-row-status status-failed"
          data-testid="bb-test-failed"
        >
          <p>{format('test_table.failed')}</p>
        </div>
      )}
      <div
        className={classNames({
          'test-table-row-category': bbTest?.method !== '',
        })}
        data-testid="bb-test-category"
      >
        {bbTest?.method === '' ? '-' : bbTest?.method}
      </div>
      <div
        className="test-table-row-endpoint"
        data-testid="bb-test-endpoint"
        data-tooltip-id={isTextOverflowing ? 'text-tooltip' : undefined}
        data-tooltip-content={isTextOverflowing ? bbTest?.endpoint : undefined}
      >
        <p ref={endpointTextContainer}>
          {bbTest?.endpoint === '' ? '-' : bbTest?.endpoint}
        </p>
        {isTextOverflowing ? (
          <TextTooltip customStyle="row-endpoint-tooltip" />
        ) : null}
      </div>
    </div>
  );
};

export default TestResultTableRow;
