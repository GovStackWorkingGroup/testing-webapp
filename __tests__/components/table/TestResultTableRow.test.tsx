import React from 'react';
import TestResultTableRow from '../../../components/table/TestResultTableRow';
import { render } from '../../test-utils/test-utils';
import { endpointTestResult } from './mockedData/ProductTable';

describe('Unit tests for TestResultTableRow component:', () => {
  const TEST_PASSED_TEST_ID = 'bb-test-passed';
  const TEST_FAILED_TEST_ID = 'bb-test-failed';
  const TEST_ENDPOINT_TEST_ID = 'bb-test-endpoint';
  const TEST_CATEGORY_TEST_ID = 'bb-test-category';

  it('should render the TestResultTableRow component with no empty fileds and match snapshot', () => {
    const { container } = render(
      <TestResultTableRow bbTest={endpointTestResult} passCurrentBBTest={jest.fn()}/>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the TestResultTableRow component with empty fields and match snapshot', () => {
    const { container } = render(
      <TestResultTableRow
        bbTest={{
          endpoint: '',
          method: '',
          passed: true,
          uri: '',
          details: []
        }}
        passCurrentBBTest={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should have corresponding values', () => {
    const { getByTestId, queryByTestId } = render(
      <TestResultTableRow bbTest={endpointTestResult} passCurrentBBTest={jest.fn()}/>
    );

    expect(queryByTestId(TEST_FAILED_TEST_ID)).toBe(null);

    expect(getByTestId(TEST_PASSED_TEST_ID)).toHaveTextContent('Passed');
    expect(getByTestId(TEST_ENDPOINT_TEST_ID)).toHaveTextContent('/data/{someparameter}/method');
    expect(getByTestId(TEST_CATEGORY_TEST_ID)).toHaveTextContent('GET');
  });
});
