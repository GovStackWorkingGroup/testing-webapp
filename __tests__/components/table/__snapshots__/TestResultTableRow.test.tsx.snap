// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for TestResultTableRow component: should render the TestResultTableRow component with empty fields and match snapshot 1`] = `
<div>
  <div
    class="test-table-row"
  >
    <div
      class="test-table-row-status status-passed"
      data-testid="bb-test-passed"
    >
      <p>
        Passed
      </p>
    </div>
    <div
      class=""
      data-testid="bb-test-category"
    >
      -
    </div>
    <div
      data-testid="bb-test-endpoint"
    >
      <p>
        -
      </p>
    </div>
  </div>
</div>
`;

exports[`Unit tests for TestResultTableRow component: should render the TestResultTableRow component with no empty fileds and match snapshot 1`] = `
<div>
  <div
    class="test-table-row"
  >
    <div
      class="test-table-row-status status-passed"
      data-testid="bb-test-passed"
    >
      <p>
        Passed
      </p>
    </div>
    <div
      class="test-table-row-category"
      data-testid="bb-test-category"
    >
      GET
    </div>
    <div
      data-testid="bb-test-endpoint"
    >
      <p>
        /data/{someparameter}/method
      </p>
    </div>
  </div>
</div>
`;
